File: Baekjoon/9663_N-Queen/B_9663.py
def n_queen(y):
    global cnt
    if y == N:
        cnt += 1
        return
    else:
        for x in range(N):
            if not vertical[x] and not diagonal1[N - (y - x) - 1] and not diagonal2[y + x]:
                vertical[x] = 1
                diagonal1[N - (y - x) - 1] = 1
                diagonal2[y + x] = 1
                n_queen(y + 1)
                vertical[x] = 0
                diagonal1[N - (y - x) - 1] = 0
                diagonal2[y + x] = 0


N = int(input())
vertical = [0] * N
diagonal1 = [0] * (2*N - 1)
diagonal2 = [0] * (2*N - 1)
cnt = 0
n_queen(0)
print(cnt)

File: Baekjoon/9663_N-Queen/B_9663.java
package baekjoon_java;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class B_9663 {
	
	private static int N;
	private static boolean[] vertical;
	private static boolean[] diagonal1;
	private static boolean[] diagonal2;
	private static int cnt = 0;
	
	private static void n_queen(int y) {
		if (y == N) {
			cnt++;
			return;
		} else {
			for (int x = 0; x < N; x++) {
				if (vertical[x] == false && diagonal1[N + x - y - 1] == false && diagonal2[x + y] == false) {
					vertical[x] = true; 
					diagonal1[N + x - y - 1] = true;
					diagonal2[x + y] = true;
					n_queen(y + 1);
					vertical[x] = false;
					diagonal1[N + x - y - 1] = false;
					diagonal2[x + y] = false;
				}
			}
		}
	}

	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		vertical = new boolean[N];
		diagonal1 = new boolean[2 * N - 1];
		diagonal2 = new boolean[2 * N - 1];
		n_queen(0);
		System.out.println(cnt);
	}

}
